schemas/schema-core/src/main/java/org/fao/geonet/kernel/schema/subtemplate/AbstractReplacer.java

```diff
@@ -0,0 +1,173 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or (at
+ * your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ *
+ * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
+ * Rome - Italy. email: geonetwork@osgeo.org
+ */
+
+package org.fao.geonet.kernel.schema.subtemplate;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.StringJoiner;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.queryparser.classic.QueryParser;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.IndexSearcher;
+import org.elasticsearch.action.search.SearchResponse;
+import org.fao.geonet.index.es.EsRestClient;
+import org.fao.geonet.kernel.schema.subtemplate.Status.Failure;
+import org.fao.geonet.utils.Xml;
+import org.jdom.Element;
+import org.jdom.JDOMException;
+import org.jdom.Namespace;
+import com.google.common.collect.ImmutableSet;
+
+public abstract class AbstractReplacer implements Replacer{
+
+    public static final float MIN_FIT_TO_REPLACE = 1f;
+
+    protected List<Namespace> namespaces;
+    protected ManagersProxy managersProxy;
+    protected ConstantsProxy constantsProxy;
+
+    public AbstractReplacer(List<Namespace> namespaces,
+                            ManagersProxy managersProxy,
+                            ConstantsProxy constantsProxy) {
+        this.constantsProxy = constantsProxy;
+        this.managersProxy = managersProxy;
+        this.namespaces = namespaces;
+    }
+
+    public Status replaceAll(Element dataXml,
+                             String localXlinkUrlPrefix,
+                             EsRestClient esRestClient,
+                             String localisedCharacterStringLanguageCode,
+                             String lang,
+                             List<String> localesAsHrefParam) {
+        List<?> nodes = null;
+        try {
+            nodes = Xml.selectNodes(dataXml, getElemXPath(), namespaces);
+        } catch (JDOMException e) {
+            return new Failure(String.format("%s- selectNodes JDOMEx: %s", getAlias(), getElemXPath()));
+        }
+        return nodes.stream()
+                .map((element) -> replace((Element) element,
+                        localXlinkUrlPrefix, esRestClient,
+                        localisedCharacterStringLanguageCode, localesAsHrefParam))
+                .collect(Status.STATUS_COLLECTOR);
+    }
+
+    protected Status replace(Element element,
```

Comment:
One thing here is that it is quite similar to existing https://github.com/geonetwork/core-geonetwork/blob/main/services/src/main/java/org/fao/geonet/api/registries/DirectoryUtils.java#L138 (not yet completely migrated to Elasticsearch). Maybe we should consider combining both approaches ?

![image](https://user-images.githubusercontent.com/1701393/140459616-c39134a3-e3b9-4571-bea9-f0dd700133d2.png)

Similarities : 
* Take a document,
* identify elements,
* check if corresponding entries are in directory,
* do something eg. here substitute as xlink

Main differences:
* DirectoryUtils is limited to one type of search and substitution at a time.
* it can work with whatever elements
* it can be played on demand like processes on uuid or a selection
* it can update the element by the directory entry AND substitute as XLink (and also create entries in the directory)
* It does exact match only
* DirectoryUtils is probably less elegant...

What do you think @cmangeat ?

Reply:
we should maybe using consider https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-similarity.html rather than querying for list of identified fields so to find the matching subtemplate.

considering the possibility of automatically creating subtemplates, it depends of the workflow, here focus was avoiding proliferation of subtemplates, letting the sole "high level" admins managing contacts dictionary for example, this way offering a service to metadata publisher: the phone number are for example assured to be up to date, and so on (office, office delegate and so on).

still wip anyway.